<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto - Part 3</font></h1>

## Introduction

In this assignment, you will be required to explore, segment, and cluster the neighborhoods in the city of Toronto based on the postalcode and borough information.. However, unlike New York, the neighborhood data is not readily available on the internet. What is interesting about the field of data science is that each project can be challenging in its unique way, so you need to learn to be agile and refine the skill to learn new libraries and tools quickly depending on the project.

For the Toronto neighborhood data, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. You will be required to scrape the Wikipedia page and wrangle the data, clean it, and then read it into a pandas  dataframe so that it is in a structured format like the New York dataset.

Once the data is in a structured format, you can replicate the analysis that we did to the New York City dataset to explore and cluster the neighborhoods in the city of Toronto.

Before we get the data and start exploring it, let's download all the dependencies that we will need.


In [3]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

#import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

!conda install -c conda-forge beautifulsoup4 --yes  # uncomment this line if you haven't completed beautifulsoup
from bs4 import BeautifulSoup as bs

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


## 1. Download and Explore Dataset

Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas  dataframe like the one shown below:


downloaded the files and placed it on the server, so you can simply run a `wget` command and access the data. So let's go ahead and do that.

In [4]:
!wget -q -O 'canada_postal_code_data.html' https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M
print('Data downloaded!')

Data downloaded!


#### Load and explore the data

In [5]:
with open("canada_postal_code_data.html") as fp:
    soup = bs(fp, 'html.parser')

tag = soup.table
body = tag.tbody

Tranform the data into dataframe


In [6]:
Tab = (body.tr).find_all('th')
colNames = [Tab[0].string.strip('\n'), Tab[1].string.strip('\n'), Tab[2].string.strip('\n')]

# instantiate the dataframe
postcodedf = pd.DataFrame(columns=colNames)

Notice how all the relevant data is in the _features_ key, which is basically a list of the neighborhoods. So, let's define a new variable that includes this data.

In [7]:
postcodedf = pd.DataFrame(columns=colNames)

# extract all 'tr' tagged fields except the first one (column names)
CodesData= body.find_all('tr')[1:]

for n, code in enumerate(CodesData): 
    tabc = ["","",""]
    for i, value in enumerate(code.stripped_strings):
        tabc[i] = value.strip()
    postcode = tabc[0]
    borough = tabc[1]
    Neighborhood = tabc[2]
    
    # Ignore borough = 'Not assigned' records 
    if borough != 'Not assigned':
        # Replace Neighborhood <= borough when Neighborhood = 'Not assigned' 
        if Neighborhood == 'Not assigned':
            Neighborhood = borough
            
        # insert the built postal code into the dataframe
        postcodedf = postcodedf.append({'PostalCode' : postcode,
                            'Borough' : borough,
                            'Neighborhood': Neighborhood},
                           ignore_index=True)

# Combine rows with same postal code into one row with the neighborhoods separated with a comma 
df = postcodedf.groupby('PostalCode', as_index=False).agg({'Borough':'first', 'Neighborhood':', '.join})
df.head(10)

PostalCode      Borough                                     Neighborhood
0        M1B  Scarborough                                   Malvern, Rouge
1        M1C  Scarborough           Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough                Guildwood, Morningside, West Hill
3        M1G  Scarborough                                           Woburn
4        M1H  Scarborough                                        Cedarbrae
5        M1J  Scarborough                              Scarborough Village
6        M1K  Scarborough      Kennedy Park, Ionview, East Birchmount Park
7        M1L  Scarborough                  Golden Mile, Clairlea, Oakridge
8        M1M  Scarborough  Cliffside, Cliffcrest, Scarborough Village West
9        M1N  Scarborough                      Birch Cliff, Cliffside West

In [8]:
print ("Toronto postal codes dataframe dimensions = ", df.shape)

Toronto postal codes dataframe dimensions =  (103, 3)


## 2. Download and Add Geographical Coordinates with Dataset

In [9]:
!wget -q -O 'GeospatialCoordinates.csv' http://cocl.us/Geospatial_data
print('Postal Code Coordinates downloaded!')

Postal Code Coordinates downloaded!


In [10]:
dfcoordinates = pd.read_csv('GeospatialCoordinates.csv')

# Rename the Postal Code column to allow merging
dfcoordinates.rename(columns={'Postal Code':'PostalCode'}, inplace=True)

#print(dfcoordinates[0:3])
#print(df[0:3]) 
df.head(10)
dflatlg = pd.merge(df, dfcoordinates, on='PostalCode' )
dflatlg

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
38         M4G         East York   
39         M4H         East York   
40         M4J         East York   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
62         M5M        North York   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A  Downtown Toronto   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353  
1               Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                    G

## 3. Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you.¶

And make sure that the dataset has all 10 boroughs and 103 neighborhoods.

In [11]:
neighborhoods = dflatlg
Torontodata = neighborhoods[neighborhoods['Borough'].str.contains('Toronto')].reset_index(drop=True)
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)
Torontodata.head(150)

The dataframe has 10 boroughs and 103 neighborhoods.


PostalCode           Borough  \
0         M4E      East Toronto   
1         M4K      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
10        M4W  Downtown Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
14        M5B  Downtown Toronto   
15        M5C  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
23        M5P   Central Toronto   
24        M5R   Central Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
27        M5V  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
30        M6G  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   
36        M6S      West Toronto   
37        M7A  Downtown Toronto   
38        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                        The Danforth West, Riverdale  43.679557 -79.352188  
2                      India Bazaar, The Beaches West  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                  North Toronto West,  Lawrence Park  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                         Moore Park, Summerhill East  43.689574 -79.383160  
9   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                        St. James Town, Cabbagetown  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                          Regent Park, Harbourfront  43.654260 -79.360636  
14                           Garden District, Ryerson  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                           Richmond, Adelaide, King  43.650571 -79.384568  
19  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
20           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                     University of Toronto, Harbord  43.662696 -79.400049  
26          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
27  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
28                                     Stn A PO Boxes  43.646435 -79.374846  
29             First Canadian Place, Underground city  43.648429 -79.382280  
30                                           Christie  43.669542 -79.422564  
31                       Dufferin, Dovercourt Village  43.669005 -79.442259  
32                          

#### Use geopy library to get the latitude and longitude values of Toronto City.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.

In [12]:
address = 'Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top.

In [13]:
# for the city Toronto, latitude and longtitude are manually extracted via google search
map_toronto = folium.Map(location = [latitude, longitude], zoom_start = 10.7)
dftoronto = Torontodata
# add markers to map
for lat, lng, borough, neighborhood in zip(dftoronto['Latitude'], dftoronto['Longitude'], dftoronto['Borough'], dftoronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_toronto)  
    

map_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

In [14]:
CLIENT_ID = 'RDKH2NBTC4WBR4GLAY3VE4LC01ECVNZBDWJL2VVUI3IPQARR' # your Foursquare ID
CLIENT_SECRET = 'JNIOHCCN3IML1NKDG0ZQPPC30LHOCZYNM4SSMFTOSYXSVLHY' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: RDKH2NBTC4WBR4GLAY3VE4LC01ECVNZBDWJL2VVUI3IPQARR
CLIENT_SECRET:JNIOHCCN3IML1NKDG0ZQPPC30LHOCZYNM4SSMFTOSYXSVLHY


#### Let's explore the first neighborhood in our dataframe.

In [15]:
Torontodata.loc[26, 'Neighborhood']

'Kensington Market, Chinatown, Grange Park'

In [16]:
Neighborhood_latitude = Torontodata.loc[26, 'Latitude']  
Neighborhood_longitude = Torontodata.loc[26, 'Longitude']  

Neighborhood_name = Torontodata.loc[26, 'Neighborhood']  

print('Latitude and longitude values of {} are {}, {}.'.format(Neighborhood_name, 
                                                               Neighborhood_latitude, 
                                                               Neighborhood_longitude))

Latitude and longitude values of Kensington Market, Chinatown, Grange Park are 43.6532057, -79.4000493.


In [17]:
#### Now, let's get the top 100 venues that are in The Beaches within a radius of 500 meters.

In [18]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    Neighborhood_latitude, 
    Neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=RDKH2NBTC4WBR4GLAY3VE4LC01ECVNZBDWJL2VVUI3IPQARR&client_secret=JNIOHCCN3IML1NKDG0ZQPPC30LHOCZYNM4SSMFTOSYXSVLHY&v=20180605&ll=43.6532057,-79.4000493&radius=500&limit=100'

Send the GET request and examine the resutls


In [19]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '6039e7b2d2329e788cd1402a'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Kensington',
  'headerFullLocation': 'Kensington, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 65,
  'suggestedBounds': {'ne': {'lat': 43.6577057045, 'lng': -79.3938414091248},
   'sw': {'lat': 43.6487056955, 'lng': -79.40625719087521}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b116957f964a520087c23e3',
       'name': 'Kid Icarus',
       'location': {'address': '205 Augusta Ave.',
        'crossStreet': 'Denison Square',
        'lat': 43.653933260442265,
        'lng': -79.40171859012935,
        'labeledLatLngs': [{'label': 'displ

let's borrow the **get_category_type** function from the Foursquare lab.

In [20]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a _pandas_ dataframe.


In [21]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name           categories        lat        lng
0                      Kid Icarus  Arts & Crafts Store  43.653933 -79.401719
1  Seven Lives - Tacos y Mariscos   Mexican Restaurant  43.654418 -79.400545
2             Blackbird Baking Co               Bakery  43.654897 -79.400619
3        Essence of Life Organics      Organic Grocery  43.654111 -79.400431
4               The Moonbean Cafe                 Café  43.654147 -79.400182

And how many venues were returned by Foursquare?


In [22]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

65 venues were returned by Foursquare.


## Explore Neighborhoods in Toronto
#### Let's create a function to repeat the same process to all the neighborhoods in Toronto


In [23]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called _Toronto_venues_.

In [24]:
Toronto_venues = getNearbyVenues(names=Torontodata['Neighborhood'],
                                   latitudes=Torontodata['Latitude'],
                                   longitudes=Torontodata['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West,  Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High

#### Let's check the size of the resulting dataframe

In [25]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1601, 7)


Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

#### Let's find out how many unique categories can be curated from all the returned venues


In [26]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 233 uniques categories.


## Analyze Each Neighborhood

In [27]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Yoga Studio  Adult Boutique  Airport  Airport Food Court  Airport Gate  \
0            0               0        0                   0             0   
1            0               0        0                   0             0   
2            0               0        0                   0             0   
3            0               0        0                   0             0   
4            0               0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       0     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Business Service  Butcher  Café  Cajun / Creole Restaurant  Candy Store  \
0                 0        0     0                          0            0   
1                 0        0     0                          0            0   
2                 0        0 

In [28]:
Toronto_onehot.shape

(1601, 233)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [29]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   Business reply mail Processing Centre, South C...     0.055556   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                  Central Bay Street     0.016393   
5                                            Christie     0.000000   
6                                Church and Wellesley     0.025316   
7                      Commerce Court, Victoria Hotel     0.000000   
8                                          Davisville     0.000000   
9                                    Davisville North     0.000000   
10                       Dufferin, Dovercourt Village     0.000000   
11             First Canadian Place, Underground city     0.000000   
12    Forest Hill North & West, Forest Hill Road Park     0.000000   
13                           Garden District, Ryerson     0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000   
15                      High Park, The Junction South     0.000000   
16                     India Bazaar, The Beaches West     0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000   
18                                      Lawrence Park     0.000000   
19                           Little Portugal, Trinity     0.023810   
20                        Moore Park, Summerhill East     0.000000   
21                 North Toronto West,  Lawrence Park     0.058824   
22                             Parkdale, Roncesvalles     0.000000   
23        Queen's Park, Ontario Provincial Government     0.033333   
24                          Regent Park, Harbourfront     0.022222   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.027027   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                     Stn A PO Boxes     0.010000   
32                                    Studio District     0.027027   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                       The Danforth West, Riverdale     0.023256   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.033333   

    Adult Boutique   Airport  Airport Food Court  Airport Gate  \
0         0.000000  0.000000            0.000000      0.000000   
1         0.000000  0.000000            0.000000      0.000000   
2         0.000000  0.000000            0.000000      0.000000   
3         0.000000  0.076923            0.076923      0.076923   
4         0.000000  0.000000            0.000000      0.000000   
5         0.000000  0.000000            0.000000      0.000000   
6         0.012658  0.000000            0.000000      0.000000   
7         0.000000  0.000000            0.000000      0.000000   
8         0.000000  0.000000            0.000000      0.000000   
9         0.000000  0.000000            0.000000      0.000000   
10        0.000000  0.000000            0.000000      0.000000   
11        0.000000  0.000000            0.000000      0.000000   
12        0.000000  0.000000            0.000000      0.000000   
13        0.000000  0.000000            0.000000      0.000000   
14        0.000000  0.000000            0.000000      0.000000   
15        0.000000  0.000000            0.000000      0.000000   
16        0.000000  0.000000            0.000000      0.000000   
17        0.000000  0.000000            0.000000    

In [30]:
Toronto_grouped.shape

(39, 233)

#### Let's print each neighborhood along with the top 5 most common venues

In [31]:
num_top_venues = 5

for hood in Toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.10
1        Cocktail Bar  0.05
2              Bakery  0.05
3            Pharmacy  0.03
4  Seafood Restaurant  0.03


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1     Coffee Shop  0.09
2          Bakery  0.09
3  Breakfast Spot  0.09
4             Bar  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.06
2       Auto Workshop  0.06
3       Garden Center  0.06
4              Garden  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0     Airport Service  0.23
1     Harbor / Marina  0.08
2             Airport  0.08
3  Airport Food Court  0.08
4        Airport Gate  0.08


----Central Bay Street----
                venue  freq
0       

#### Let's put that into a _pandas_ dataframe

First, let's write a function to sort the venues in descending order.

In [32]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [33]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Cocktail Bar        Farmers Market   
1           Coffee Shop                Bakery        Breakfast Spot   
2           Yoga Studio         Auto Workshop         Garden Center   
3      Sculpture Garden        Airport Lounge   Rental Car Location   
4        Sandwich Place                  Café    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar    Seafood Restaurant           Cheese Shop   
1     Convenience Store          Intersection    Italian Restaurant   
2                Garden  Gym / Fitness Center  Fast Food Restaurant   
3       Harbor / Marina      Airport Terminal         Boat or Ferry   
4           Salad Place   Japanese Restaurant       Thai Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Pharmacy            Restaurant           Gourmet Shop  
1            Restaurant          Climbing Gym                    Bar  
2        Farmers Market            Comic Shop            Pizza Place  
3          Airport Gate    Airport Food Court                Airport  
4       Bubble Tea Shop          Burger Joint            Yoga Studio

## Cluster Neighborhoods

Run _k_-means to cluster the neighborhood into 5 clusters.


In [34]:
# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 3, 3, 3, 3, 3, 3, 3, 3, 3], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [35]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = Torontodata

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head() # check the last columns!

PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0                 Trail     Health Food Store   
1 -79.352188               3      Greek Restaurant           Coffee Shop   
2 -79.315572               3  Fast Food Restaurant           Pizza Place   
3 -79.340923               3           Coffee Shop             Gastropub   
4 -79.388790               3                  Park      Business Service   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0                   Pub             Wine Shop            Dance Studio   
1    Italian Restaurant            Restaurant  Furniture / Home Store   
2      Sushi Restaurant     Food & Drink Shop          Sandwich Place   
3   American Restaurant               Brewery                    Café   
4              Bus Line           Swim School       Electronics Store   

         6th Most Common Venue 7th Most Common Venue  \
0  Eastern European Restaurant   Dumpling Restaurant   
1               Ice Cream Shop                  Café   
2                      Brewery          Liquor Store   
3                       Bakery    Italian Restaurant   
4  Eastern European Restaurant   Dumpling Restaurant   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Donut Shop      Doner Restaurant                Dog Run  
1                     Bakery  Caribbean Restaurant                    Spa  
2         Italian Restaurant            Restaurant         Ice Cream Shop  
3  Latin American Restaurant             Bookstore     Seafood Restaurant  
4                 Donut Shop      Doner Restaurant                Dog Run

In [36]:
neighborhoods_venues_sorted.head()

Cluster Labels                                       Neighborhood  \
0               3                                        Berczy Park   
1               3       Brockton, Parkdale Village, Exhibition Place   
2               3  Business reply mail Processing Centre, South C...   
3               3  CN Tower, King and Spadina, Railway Lands, Har...   
4               3                                 Central Bay Street   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop                Bakery          Cocktail Bar   
1                  Café           Coffee Shop                Bakery   
2    Light Rail Station           Yoga Studio         Auto Workshop   
3       Airport Service      Sculpture Garden        Airport Lounge   
4           Coffee Shop        Sandwich Place                  Café   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0        Farmers Market              Beer Bar    Seafood Restaurant   
1        Breakfast Spot     Convenience Store          Intersection   
2         Garden Center                Garden  Gym / Fitness Center   
3   Rental Car Location       Harbor / Marina      Airport Terminal   
4    Italian Restaurant           Salad Place   Japanese Restaurant   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0           Cheese Shop              Pharmacy            Restaurant   
1    Italian Restaurant            Restaurant          Climbing Gym   
2  Fast Food Restaurant        Farmers Market            Comic Shop   
3         Boat or Ferry          Airport Gate    Airport Food Court   
4       Thai Restaurant       Bubble Tea Shop          Burger Joint   

  10th Most Common Venue  
0           Gourmet Shop  
1                    Bar  
2            Pizza Place  
3                Airport  
4            Yoga Studio

Finally, let's visualize the resulting clusters

Finally, let's visualize the resulting clusters


In [38]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

Now,  examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories,  assign a name to each cluster.

#### Cluster 1


In [39]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  East Toronto               0                 Trail     Health Food Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub             Wine Shop          Dance Studio   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant                Dog Run

#### Cluster 2

In [40]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8  Central Toronto               1                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8             Wine Shop          Dance Studio     Electronics Store   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
8      Doner Restaurant               Dog Run    Distribution Center

#### Cluster 3

In [41]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               2  Fast Food Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22                Garden             Wine Shop         Deli / Bodega   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
22     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22            Donut Shop      Doner Restaurant                Dog Run

#### Cluster 4

In [42]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1       East Toronto               3      Greek Restaurant   
2       East Toronto               3  Fast Food Restaurant   
3       East Toronto               3           Coffee Shop   
4    Central Toronto               3                  Park   
5    Central Toronto               3                  Park   
6    Central Toronto               3           Coffee Shop   
7    Central Toronto               3           Pizza Place   
9    Central Toronto               3           Coffee Shop   
11  Downtown Toronto               3           Coffee Shop   
12  Downtown Toronto               3           Coffee Shop   
13  Downtown Toronto               3           Coffee Shop   
14  Downtown Toronto               3           Coffee Shop   
15  Downtown Toronto               3                  Café   
16  Downtown Toronto               3           Coffee Shop   
17  Downtown Toronto               3           Coffee Shop   
18  Downtown Toronto               3           Coffee Shop   
19  Downtown Toronto               3           Coffee Shop   
20  Downtown Toronto               3           Coffee Shop   
21  Downtown Toronto               3           Coffee Shop   
24   Central Toronto               3        Sandwich Place   
25  Downtown Toronto               3                  Café   
26  Downtown Toronto               3                  Café   
27  Downtown Toronto               3       Airport Service   
28  Downtown Toronto               3           Coffee Shop   
29  Downtown Toronto               3           Coffee Shop   
30  Downtown Toronto               3         Grocery Store   
31      West Toronto               3                Bakery   
32      West Toronto               3                   Bar   
33      West Toronto               3                  Café   
34      West Toronto               3                  Park   
35      West Toronto               3             Gift Shop   
36      West Toronto               3           Coffee Shop   
37  Downtown Toronto               3           Coffee Shop   
38      East Toronto               3    Light Rail Station   

            2nd Most Common Venue  3rd Most Common Venue  \
1                     Coffee Shop     Italian Restaurant   
2                     Pizza Place       Sushi Restaurant   
3                       Gastropub    American Restaurant   
4                Business Service               Bus Line   
5                    Dance Studio         Breakfast Spot   
6                  Clothing Store    Sporting Goods Shop   
7                    Dessert Shop         Sandwich Place   
9                Sushi Restaurant                    Pub   
11                    Pizza Place              Pet Store   
12            Japanese Restaurant       Sushi Restaurant   
13                           Park                 Bakery   
14                 Clothing Store        Bubble Tea Shop   
15                    Coffee Shop    American Restaurant   
16                         Bakery           Cocktail Bar   
17                 Sandwich Place                   Café   
18                           Café             Restaurant   
19                       Aquarium                  Hotel   
20                          Hotel                   Café   
21                     Restaurant                   Café   
24                           Café            Coffee Shop   
25            Japanese Restaurant                    Bar   
26  Vegetarian / Vegan Restaurant            Coffee Shop   
27               Sculpture Garden         Airport Lounge   
28             Seafood Restaurant           Cocktail Bar   
29                           Café                  Hotel   
30                           Café                   Park   
31                       Pharmacy              Wine Shop   
32                    Men's Store  Vietnamese Restaurant   
33                    Coffee Shop                 Bakery   
34             Mexican Restaurant         

#### Cluster 5

In [43]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
10  Downtown Toronto               4                  Park   
23   Central Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10            Playground                 Trail          Cupcake Shop   
23         Jewelry Store                 Trail      Sushi Restaurant   

          5th Most Common Venue 6th Most Common Venue  \
10  Eastern European Restaurant   Dumpling Restaurant   
23                    Wine Shop         Deli / Bodega   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
10                   Donut Shop      Doner Restaurant               Dog Run   
23  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   10th Most Common Venue  
10    Distribution Center  
23       Doner Restaurant

### Thank you